### PR TITLE
Thread safety of printing and matmul

### LINKED ISSUE #442: Matrix multiplication not thread safe.
MRE: several copies of a matrix are multiplied by the same matrix several times independently. The result should be the same for all copies, but the function usually returns false for N>=4. Some multiplication are not performed correctly.

```
function testMM(T, R, N)
    m0 = rand(0:2, (N,N))
    m1 = rand(0:2, (N,N))
    m = [deepcopy(m0) for i=1:R]
    Threads.@threads for i=1:R
        for t=1:T
            m[i] *= m1
        end
    end
    return all(x->(x==m[1]),m)
end 

testMM(5,8,3)
# true

testMM(5,8,4)
# false - most likely
```

`Threads.nthreads()` is 4 for me. With a normal For-loop without threads the function always returns true.

The issue only happens with matrix multiplication, but not with addition, multiplying by a scalar or element-wise multiplication.

```
Julia Version 0.6.0
Commit 903644385b* (2017-06-19 13:05 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
```

### LINKED ISSUE #21335: (FloatingPoint -> String),sprintf not thread safe
They both use the global DIGITS buffer, so the threads get parts of each other's numbers.

### LINKED ISSUE #22581: Matrix multiplication not thread safe.
MRE: several copies of a matrix are multiplied by the same matrix several times independently. The result should be the same for all copies, but the function usually returns false for N>=4. Some multiplication are not performed correctly.

```
function testMM(T, R, N)
    m0 = rand(0:2, (N,N))
    m1 = rand(0:2, (N,N))
    m = [deepcopy(m0) for i=1:R]
    Threads.@threads for i=1:R
        for t=1:T
            m[i] *= m1
        end
    end
    return all(x->(x==m[1]),m)
end 

testMM(5,8,3)
# true

testMM(5,8,4)
# false - most likely
```

`Threads.nthreads()` is 4 for me. With a normal For-loop without threads the function always returns true.

The issue only happens with matrix multiplication, but not with addition, multiplying by a scalar or element-wise multiplication.

```
Julia Version 0.6.0
Commit 903644385b* (2017-06-19 13:05 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Sandybridge)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, ivybridge)
```

### LINKED ISSUE #25727: Grisu (floating point printing) not thread-safe
It looks like string interpolation is currently not threadsafe. Running the following script "string_interpolation_threadsafe_test.jl"

```julia
using Base.Threads


function test_with_mutex()
	a = rand(1000)
	m = Mutex()
	@threads for i=1:1000
		y = lock(m) do
			"$(a[i])"
		end
		@assert a[i] == parse(Float64, y)
	end
end


function test_without_mutex()
	a = rand(1000)
	@threads for i=1:1000
		y = "$(a[i])"
		@assert a[i] == parse(Float64, y)
	end
end


function test_without_mutex_repeated()
	for i in 1:1000
		a = rand(1000)
		@threads for i=1:1000
			y = "$(a[i])"
		end
	end
end


@info "test_with_mutex()"
test_with_mutex()

@info "test_without_mutex()"
test_without_mutex()

@info "test_without_mutex_repeated()"
test_without_mutex_repeated()
```

via

```shell
JULIA_NUM_THREADS=4 julia string_interpolation_threadsafe_test.jl
```

with Julia 0.7.0-DEV.3551 (x86_64, official binary, on Ubuntu 16.04)  results in

```
[ Info: test_with_mutex()
[ Info: test_without_mutex()

Error thrown in threaded loop on thread 1: AssertionError(msg="a[i] == parse(Float64, y)")
Error thrown in threaded loop on thread 0: AssertionError(msg="a[i] == parse(Float64, y)")
Error thrown in threaded loop on thread 3: AssertionError(msg="a[i] == parse(Float64, y)")
Error thrown in threaded loop on thread 2: AssertionError(msg="a[i] == parse(Float64, y)")

[ Info: test_without_mutex_repeated()

Error thrown in threaded loop on thread 0: InexactError(func=:trunc, T=UInt8, val=-1)
Error thrown in threaded loop on thread 3: DivideError()
Error thrown in threaded loop on thread 0: DivideError()
Error thrown in threaded loop on thread 1: DivideError()
Error thrown in threaded loop on thread 0: InexactError(func=:trunc, T=UInt8, val=-1)
Error thrown in threaded loop on thread 1: InexactError(func=:trunc, T=UInt8, val=-1)
```

`test_without_mutex_repeated()` sometimes ends in a segfault.

The same problem occurs with Julia v0.6.

As string interpolation may be used deep inside a third-party function/dependency, and is therefore hard to control/fence, this is a serious problem when writing multi-threaded code.

### PR DESCRIPTION
**Pull request by stevengj**
This PR uses a per-thread buffer to make the generic-matmul thread-safe (fixes JuliaLang/LinearAlgebra.jl#442), floating-point printing thread-safe (fixes JuliaLang/julia#25727), and printf thread-safe (fixes JuliaLang/julia#21335).

In particular, the old `DIGITS` array in grisu is supplemented by a `DIGITSs = [DIGITS]` array that is expanded by `__init__` into a per-thread buffer.  I kept the old `DIGITS` array to avoid breaking the few external packages that use this — they can migrate gradually to the thread-safe version.

The basic matmul buffer structure is not changed, so JuliaLang/LinearAlgebra.jl#500 is neither better nor worse.

As I mentioned in JuliaLang/julia#10441, it would be good to have better support for thread-local variables in the compiler.  Ideally I could just put `@threadlocal const DIGITS=[....]` and it would do the right thing under the hood.

### CODE REVIEW

**Comment by Keno** [comment]
This approach seems fine with me for now. We'll have to carefully think this through with the new threading runtime, since that'll start scheduling tasks on different threads, so thread-local will no longer be right. cc @kpamnany

**Comment by JeffBezanson** [comment]
Right, in general we need task-local storage, but I expect thread-local to be much more efficient (since there's a fixed number of threads, among other things). So we should continue to use this technique for internal buffers where we know that tasks won't block while using them (they won't, right? right? :smile: ).

**Comment by Keno** [comment]
One possible model would be to an explicit acquire/release for task local buffers, such that they get automatically copied along with the task if switching does happen.

**Comment by stevengj** [comment]
CI failures are unrelated, I think.  MacOS Travis is failing with `fatal error: 'cholmod.h' file not found`, and 32-bit appveyor seems to be a timeout.

**Comment by JeffBezanson** [comment]
Would be good to spot check that float to string conversion and printf performance are unaffected. I expect there would be no difference, but you never know.

**Comment by stevengj** [comment]
old (`master`) thread-unsafe version:
```jl
julia> foo(x) = @sprintf("%3.4f %10.3g", x, x);
          foo(io, x) = @printf(io, "%3.4f %10.3g", x, x);

julia> @btime string(3.7234);
  338.630 ns (4 allocations: 192 bytes)

julia> @btime foo(3.7234);
  738.968 ns (10 allocations: 576 bytes)

julia> buf = IOBuffer(sizehint=100); @btime print($buf, 3.7234);
  255.677 ns (0 allocations: 0 bytes)

julia> buf = IOBuffer(sizehint=100); @btime foo($buf, 3.7234);
  717.572 ns (5 allocations: 288 bytes)
```
new thread-safe version:
```jl
julia> @btime string(3.7234);
  332.072 ns (4 allocations: 192 bytes)

julia> @btime foo(3.7234);
  760.429 ns (10 allocations: 576 bytes)

julia> buf = IOBuffer(sizehint=100); @btime print($buf, 3.7234);
  257.676 ns (0 allocations: 0 bytes)

julia> buf = IOBuffer(sizehint=100); @btime foo($buf, 3.7234);
  745.630 ns (5 allocations: 288 bytes)
```
So there seems to be at most a 4% slowdown, usually less.  This seems a worthwhile price to pay for thread safety.

**Comment by stevengj** [comment]
Okay to merge?

**Review by Keno — APPROVED** [review]

File: base/grisu/grisu.jl [hunk 1]
```diff
@@ -3,14 +3,20 @@
...
 const SHORTEST = 1
 const FIXED = 2
 const PRECISION = 3
 
 const DIGITS = Vector{UInt8}(undef, 309+17)
 
+# thread-safe code should use a per-thread DIGITS buffer DIGITSs[Threads.threadid()]
+const DIGITSs = [DIGITS]
+function __init__()
+    Threads.resize_nthreads!(DIGITSs)
```
Comment (musm) [inline base/grisu/grisu.jl]: Is this supposed to be   `resize!(DIGITSs, Threads.nthreads())` ?

**Comment by vtjnash** [comment]
> access to DIGITSs is inbounds

I think I would prefer we drop this commit unless it is absolutely conclusive that it is essential. We can refactor some of this code to move the computation of the buffer outside hot loops if necessary.

**Comment by stevengj** [comment]
@vtjnash, reverted.  I just tested and it makes no measurable difference in performance.

**Comment by stevengj** [comment]
Unrelated timeout error on AppVeyor 64-bit

**Comment by simonbyrne** [comment]
Do we need to do the same thing for `Grisu.BIGNUMS` as well?

@quinnj do you know?

**Comment by quinnj** [comment]
Hmmmm, not sure. I think I remember @vtjnash fixing that a while ago though.

**Comment by simonbyrne** [comment]
No, I managed to trigger an issue. See #27826

**Pull request merged**